llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1392,7 +1395,19 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
                                                   TypeInfo.inferScalarType(Sel),
                                                   false, false, false);
               })
-
+              .Case<VPInstruction>([&](VPInstruction *VPI) -> VPRecipeBase * {
+                VPValue *LHS, *RHS;
+                if (!match(VPI, m_Select(m_Specific(HeaderMask), m_VPValue(LHS),
```

Comment:
Would be good to document the code here

Do we need this for any select, or just the ones for out-loop reductions?

Reply:
doc it dc2da769fabe791feee128e4e401e55301c4881a
As far as I know, currently only out-loop reduction requires this transformation.

Reply:
right, what I meant was that there could be other selects and we are not checking for reductions here?

Reply:
I understand your concern. However, I don’t think we need to perform a check specifically for reductions. As long as the `VPInstruction::select` matches the form `select(HeaderMask, LHS, RHS)`, it is correct to convert it to `vp.merge(all-true, LHS, RHS, EVL)`, whether or not it’s a predicated reduction select.

The difference between `vp.merge` and `vp.select` is that for all result lanes at positions greater or equal than EVL, `vp.select` sets them as  undefined, whereas `vp.merge` sets those lanes to `RHS` (the value on false). Thus, `VPWidenSelectRecipe` and `VPInstruction::select` **without** a header mask condition can be converted to `vp.select` because we don't care the results at inactive lanes, while `VPInstruction::select` **with** a header mask condition must be converted to `vp.merge`.

Reply:
Sounds good, but then the comment should be updated to not refer to predicted reductions I think

Reply:
53d9ac6b0aa6d890e82cdffd53cd61faa025f772
Agree, replace `predicated reduction select` with `select with a header mask condition`.